plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/state/MapboxFollowingState.kt

```diff
@@ -0,0 +1,149 @@
+package com.mapbox.maps.plugin.viewport.experimental.state
+
+import android.animation.AnimatorSet
+import android.animation.ValueAnimator
+import androidx.core.animation.doOnEnd
+import com.mapbox.geojson.Point
+import com.mapbox.maps.CameraOptions
+import com.mapbox.maps.plugin.animation.camera
+import com.mapbox.maps.plugin.delegates.MapDelegateProvider
+import com.mapbox.maps.plugin.locationcomponent.OnIndicatorBearingChangedListener
+import com.mapbox.maps.plugin.locationcomponent.OnIndicatorPositionChangedListener
+import com.mapbox.maps.plugin.locationcomponent.location
+import com.mapbox.maps.plugin.viewport.ViewportPlugin
+import com.mapbox.maps.plugin.viewport.experimental.Cancelable
+import com.mapbox.maps.plugin.viewport.experimental.data.FollowingViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.transition.MapboxViewportTransitionFactory
+import java.util.concurrent.*
+
+/**
+ * Describes state when [ViewportPlugin] finished transition to the following state.
+ *
+ * Set after invoking [ViewportPlugin.requestCameraToFollowing].
+ *
+ * When in this state, each update to the [ViewportDataSource]
+ * will automatically trigger another transition.
+ */
+class MapboxFollowingState internal constructor(
+  mapDelegateProvider: MapDelegateProvider,
+  override var options: FollowingViewportStateOptions
+) : FollowingViewportState {
+  private val cameraPlugin = mapDelegateProvider.mapPluginProviderDelegate.camera
+  private val locationComponent = mapDelegateProvider.mapPluginProviderDelegate.location
+  private val transitionFactory = MapboxViewportTransitionFactory(mapDelegateProvider)
+  private val dataSourceUpdateObservers = CopyOnWriteArrayList<ViewportDataSourceObserver>()
+
+  private var lastLocation: Point = Point.fromLngLat(0.0, 0.0)
+  private var lastBearing: Double = 0.0
+
+  private var runningAnimation: AnimatorSet? = null
+  private var isFollowingStateRunning = false
+  private var isObservingLocationUpdates = false
+
+  private val indicatorPositionChangedListener = OnIndicatorPositionChangedListener { point ->
+    lastLocation = point
+    notifyLatestViewportData()
+  }
+  private val indicatorBearingChangedListener = OnIndicatorBearingChangedListener { bearing ->
+    lastBearing = bearing
+    notifyLatestViewportData()
+  }
+
+  private fun notifyLatestViewportData() {
+    val viewportData = evaluateViewportData()
+    if (isFollowingStateRunning) {
+      // Use instant update here since the location updates are already interpolated by the location component plugin
+      updateFrame(viewportData, true)
+    }
+    dataSourceUpdateObservers.forEach {
+      if (!it.onNewData(viewportData)) {
+        dataSourceUpdateObservers.remove(it)
+      }
+    }
+  }
+
+  private fun evaluateViewportData(): CameraOptions {
+    // TODO implement the complete algo to calculate the viewport data according to data source options.
```

Comment:
perhaps link a task here?

Reply:
With current design, we provide a very basic tracking modes that only relies on the location puck's position update, and kept the minimum config options. We will add more config options and adjust the algo later on if there's request, will remove this TODO.